### PR TITLE
feat: alex swap integration (alex team)

### LINKED ISSUE #4158: Integrate ALEX swaps
Creating this issue to track helping to broadcast swap transactions.

### PR DESCRIPTION
**Pull request by zhigang1992**
This is an early version of the swap integration.

We have the alex-sdk connected and all the readonly functions working.

We need help actually broadcast the tx.

### CODE REVIEW

File: src/app/pages/swap/swap-container.tsx [hunk 1]
```diff
@@ -1,66 +1,128 @@
...
+    );
+    const txToBroadcast = alexSDK.runSwap(
+      '', // TODO: current user's stxAddress
+      swapSubmissionData.swapAssetFrom.currency,
+      swapSubmissionData.swapAssetTo.currency,
+      fromAmount,
+      minToAmount,
+      swapSubmissionData.router.map(x => x.currency)
+    );
+    // TODO: broadcast the tx
```
Comment (zhigang1992) [inline src/app/pages/swap/swap-container.tsx]: Here are two TODO

The `txToBroadcast` contains everything needed to assemble the StackTransaction.

```typescript
export type TxToBroadCast = {
  contractAddress: string;
  contractName: string;
  functionName: string;
  functionArgs: ClarityValue[];
  postConditions: Array<FungiblePostCondition | STXPostCondition>;
};
```

Please help point me at the right place to
1. Read current user's stx address
2. Broadcast this tx.

File: src/app/pages/swap/swap-container.tsx [hunk 2]
```diff
@@ -1,66 +1,128 @@
...
-import { useNativeSegwitBalance } from '@app/query/bitcoin/balance/bitcoin-balances.query';
-import { useCurrentAccountNativeSegwitIndexZeroSigner } from '@app/store/accounts/blockchain/bitcoin/native-segwit-account.hooks';
 
 import { SwapContainerLayout } from './components/swap-container.layout';
 import { SwapForm } from './components/swap-form';
 import { SwapAsset, SwapFormValues } from './hooks/use-swap';
-import { SwapContext, SwapProvider } from './swap.context';
-
-// TODO: Remove and set to initial state to 0 with live data
-const tempExchangeRate = 0.5;
```
Comment (zhigang1992) [inline src/app/pages/swap/swap-container.tsx:21]: The exchange rate is not a constant, we have to call readonly function for each amount that user input.

We removed this field and added a function to fetch the toAmount.

File: src/app/pages/swap/swap-container.tsx [hunk 3]
```diff
@@ -1,66 +1,128 @@
...
+  const swappableAssets: SwapAsset[] = useMemo(
+    () => supportedCurrencies.map(getAssetFromAlexCurrency),
+    [getAssetFromAlexCurrency, supportedCurrencies]
+  );
 
-  function onSubmitSwapForReview(values: SwapFormValues) {
-    navigate(RouteUrls.SwapReview, {
-      state: { ...values },
+  const [swapSubmissionData, setSwapSubmissionData] = useState<SwapSubmissionData>();
+  const [slippage, _setSlippage] = useState(0.04);
```
Comment (zhigang1992) [inline src/app/pages/swap/swap-container.tsx:67]: Currently the UI is missing the ability to update slippage, we have it default to 4% atm.

File: src/app/pages/swap/components/swap-details/swap-details.tsx [hunk 4]
```diff
@@ -1,14 +1,34 @@
+import { useSwapContext } from '@app/pages/swap/swap.context';
+
 import { SwapDetailLayout } from './swap-detail.layout';
 import { SwapDetailsLayout } from './swap-details.layout';
 
-// TODO: Replace with live data
 export function SwapDetails() {
+  const { swapSubmissionData } = useSwapContext();
+  if (swapSubmissionData == null) {
```
Comment (kyranjamie) [inline src/app/pages/swap/components/swap-details/swap-details.tsx]: We have a `isDefined` helper that can avoid `==` operators

File: src/app/pages/swap/swap-container.tsx [hunk 5]
```diff
@@ -1,66 +1,128 @@
...
 export function SwapContainer() {
-  const [exchangeRate, setExchangeRate] = useState(tempExchangeRate);
-  const [isSendingMax, setIsSendingMax] = useState(false);
+  const alexSDK = useState(() => new AlexSDK())[0];
+  const [supportedCurrencies, setSupportedCurrencies] = useState<TokenInfo[]>([]);
+
+  useEffect(() => {
+    alexSDK.fetchTokenList().then(tokenList => {
+      setSupportedCurrencies(tokenList.filter(t => t.availableInSwap));
+    });
```
Comment (kyranjamie) [inline src/app/pages/swap/swap-container.tsx]: Could this be represented as a React Query hook?

File: src/app/pages/swap/swap-container.tsx [hunk 6]
```diff
@@ -1,66 +1,128 @@
...
   }
 
+  async function fetchToAmount(
+    from: SwapAsset,
+    to: SwapAsset,
+    fromAmount: string
+  ): Promise<string> {
+    const result = await alexSDK.getAmountTo(
+      from.currency,
+      BigInt(new BigNumber(fromAmount).multipliedBy(1e8).dp(0).toString()),
```
Comment (kyranjamie) [inline src/app/pages/swap/swap-container.tsx]: not sure what `1e8` is, can we make a named variable to describe it?

**Review by kyranjamie — COMMENTED** [review]
Thanks for the PR

File: src/app/pages/swap/swap-container.tsx (hunk 6 above)
Comment (fiftyeightandeight) [reply src/app/pages/swap/swap-container.tsx]: Hi @kyranjamie we use 8-digit fixed notation to represent decimals, hence multiplying a float by 1e8 and casting to `BigInt`.

Comment (kyranjamie) [reply src/app/pages/swap/swap-container.tsx]: Got it, would prefer to avoid magic numbers though

```ts
const oneHundredMillion = 100_000_000;
```

We also have a `stxToMicroStx` helper too

File: src/app/pages/swap/swap-container.tsx (hunk 2 above)
Comment (fbwoolf) [reply src/app/pages/swap/swap-container.tsx:21]: Yep, figured this would be removed, thx.

File: src/app/pages/swap/components/swap-selected-asset-from.tsx [hunk 7]
```diff
@@ -18,14 +18,12 @@ const maxAvailableCaption = 'Max available in your balance';
 const maxAvailableTooltip =
   'Amount of funds that is immediately available for use, after taking into account any pending transactions or holds placed on your account by the protocol.';
 
-const sendAnyValue = 'Send any value';
```
Comment (fbwoolf) [inline src/app/pages/swap/components/swap-selected-asset-from.tsx]: Why was this changed? Can you provide a video or screenshot of the UI and tag @mica000 to approve the design change?

File: src/app/pages/swap/swap-container.tsx (hunk 1 above)
Comment (fbwoolf) [reply src/app/pages/swap/swap-container.tsx]: You can use `useCurrentAccountStxAddressState()` for the address. Does `useTransactionBroadcast()` help?

File: src/app/pages/swap/components/swap-selected-asset-from.tsx (hunk 7 above)
Comment (zhigang1992) [reply src/app/pages/swap/components/swap-selected-asset-from.tsx]: Sry, this was not intentional. Will revert.

File: src/app/pages/swap/swap.context.ts [hunk 8]
```diff
@@ -2,11 +2,19 @@ import { createContext, useContext } from 'react';
...
+  liquidityFee: number;
+  router: SwapAsset[];
+  slippage: number;
+}
+
 export interface SwapContext {
-  exchangeRate: number;
-  isSendingMax: boolean;
-  onSetExchangeRate(value: number): void;
-  onSetIsSendingMax(value: boolean): void;
```
Comment (fbwoolf) [inline src/app/pages/swap/swap.context.ts:9]: Working on this locally, curious why you removed all the code to set the send max? It seemed like a good way to toggle that imo and utilize the state throughout the feature? I think we want to keep it bc others will be using this code too.

**Comment by markmhendrickson** [comment]
@fbwoolf is continuing work here on https://github.com/leather-wallet/extension/pull/4205

**Comment by kyranjamie** [comment]
@fbwoolf can we close now?

**Comment by fbwoolf** [comment]
Yep, we can close this. Will do.
